Export: Populate advanced settings menus with API data

jordwest opened the pull request:
Populates the Exporter advanced settings dropdown menus with data fetched from the API:

<img width="570" alt="screen shot 2016-02-18 at 6 46 34 pm" src="https://cloud.githubusercontent.com/assets/416133/13138237/a7ac690e-d671-11e5-8163-a641a048da3d.png">

The `OptionFieldset` component has also been renamed `PostTypeOptions` to better describe its purpose.

### Testing

0. ~~Apply patch D1816 and sandbox public-api.wordpress.com~~ Deployed in r135999-wpcom
1. Go to `/settings/export` and choose a site.
2. Click the down chevron beside **Export All**.
3. Confirm that the author, status, date and category options under each post type belong to the site you've chosen.
4. Switch sites.
5. Confirm that the options in the dropdowns have changed and match the new site.

cc @dllh @omarjackman @dmsnell

dllh commented on client/my-sites/exporter/post-type-options.jsx:
```diff
@@ -0,0 +1,174 @@
+		return (
+			<div className="exporter__option-fieldset-fields">
+				{ options.authors &&
+					<Select defaultLabel="Author…" options={ mapAuthors( options.authors ) } { ...commonProps } />
```
These default labels should be translated, I think?

dllh commented:
The options update as expected. When I click the "Export Selected Content" button, the request goes to `https://public-api.wordpress.com/rest/v1.1/sites/[object%20Object]/exports/start?http_envelope=1` and nothing happens, but I think that's probably more or less expected and will be fixed up in a later PR.

This looks good to me with that and the translation thing noted. @omarjackman, please take a look and if all looks good to you as well, let's :ship:.

omarjackman commented on client/my-sites/exporter/post-type-options.jsx:
```diff
@@ -0,0 +1,174 @@
+				value: status.name,
+				label: status.label
+			} ) );
+		}
```
semi-colon :smile:

omarjackman commented on client/my-sites/exporter/post-type-options.jsx:
```diff
@@ -0,0 +1,174 @@
+				value: category.name,
+				label: category.name
+			} ) );
+		}
```
semi-colon :smile:

omarjackman commented on client/my-sites/exporter/post-type-options.jsx:
```diff
@@ -0,0 +1,174 @@
+					label: time.format( 'MMM YYYY' )
+				}
+			} );
+		}
```
semi-colon :smile:

omarjackman commented on client/my-sites/exporter/select.jsx:
```diff
@@ -0,0 +1,24 @@
+
+		return (
+			<FormSelect { ...this.props }>
+				<option key="default" value="">{ this.props.defaultLabel }</option>
```
I don't think you need `key` here since your not iterating.

omarjackman commented on client/state/site-settings/exporter/selectors.js:
```diff
@@ -36,5 +36,11 @@ export function isExporting( state, siteId ) {
-export const advancedSettings = ( state, siteId ) => state.siteSettings.exporter.advancedSettings[ siteId ];
+export const getPostTypeOptions = ( state, siteId, postType ) => {
+	const advancedSettings = getAdvancedSettings( state, siteId );
+	if ( ! advancedSettings ) return null;
```
We should always use curly braces here but I think you'd get away with a one line return like 
`return advancedSettings ? advancedSettings[ postType ] : null;`

omarjackman commented on client/my-sites/exporter/post-type-options.jsx:
```diff
@@ -0,0 +1,174 @@
+		const commonProps = {
+			disabled: ! this.props.isEnabled
+		};
+
```
Just a suggestion:

I see a lot of repeated `value`/`label` and also function declarations that are only used once. Maybe you could use one function and simplify all of the other mappings like this
```js
const postTypeOption = ( value, label ) => { value, label };
const mappedAuthors = options.authors && options.authors.map( author => postTypeOption( author.ID, author.name ) );
const mappedStatuses = options.statuses && options.statuses.map( status => postTypeOption( status.name, status.label ) );
...
```

And then...

```js
...
{ mappedAuthors && <Select defaultLabel="Status…" options={ mappedAuthors } { ...commonProps } /> }
...
```

You could probably take it even further to eliminate the repeating with

```js
const selectProps = [];
options.authors && selectProps.push( {
    defaultLabel: this.translate( 'Author…' ),
    options: options.authors.map( author => postTypeOption( author.ID, author.name ) )
} );
...
```

then

```js
...
{ selectProps.map( props => <Select { ...props } disabled={ disabled } /> }
...
```

This is just a suggestion and not trying to rewrite :smile:

jordwest commented:
Thanks for the feedback @omarjackman, I'm finally jumping back into this. Would you mind taking another look? The first commit was rebased, but all other changes since you reviewed are in the last two commits.

I've just moved a bunch of unnecessary logic out of the client and handled it the server. This addresses the feedback above, specifically [this one](https://github.com/Automattic/wp-calypso/pull/3359#discussion-diff-53419700). Patch D1816 is now needed for this branch to work.

@dllh:
> nothing happens, but I think that's probably more or less expected and will be fixed up in a later PR.

Right, this PR won't do the actual request - that'll come in the next PR.

dmsnell commented:
I'm getting some blank fields on Chrome Mac. The status dropdown is filled out, but the others appear to have blanks.

<img width="564" alt="screen shot 2016-05-16 at 12 47 45 pm" src="https://cloud.githubusercontent.com/assets/5431237/15296713/9a6fc8ac-1b64-11e6-8ca0-356da44b734b.png">

jordwest commented:
@dmsnell oops sorry, I hadn't updated the original testing instructions with the patch. Could you apply D1816 and sandbox public-api.wordpress.com and try again?

dllh commented on client/my-sites/exporter/post-type-options.jsx:
```diff
@@ -0,0 +1,129 @@
+
+		const Field = ( props ) => {
+			if ( ! props.options ) {
+				// This should be replaced with `return null` in React >= 0.15
```
Is this something we'll have to go back and remember to update?

jordwest replied:
The current code will continue to work in React >= 0.15, it's just that `null` makes more sense than an empty `<span/>` once it's [supported in React 15](https://facebook.github.io/react/blog/#functional-components-can-now-return-null-too)

dllh commented on client/my-sites/exporter/post-type-options.jsx:
```diff
@@ -0,0 +1,129 @@
+
+		return (
+			<div className="exporter__option-fieldset-fields">
+				<Field defaultLabel="Author…" options={ options.authors } />
```
These should be translated, right?

dllh commented:
Just refreshed and tested, and it works well for me. The code looks good at a glance too, with a couple of minor questions.

jordwest commented:
Thanks for the review. The changes in the server patch have been committed now, so no more sandboxing needed. If there are no other objections I'll merge this evening.
